File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/scala/test_all.scala
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

package test

trait Result[T] {
  def get_result(): T
}

object TestAll {

def run_all_tests[Args <: Result[Ret], Ret](name: String, args_list: Array[Args], fnc: (Args) => Ret): Int =
{
    var nb_errors: Int = 0
    printf("\n" + name + "\n")
    var nb: Int = 1
    for (args <- args_list)
    {
        val res = fnc(args)
        val expected = args.get_result()
        var is_same = false
        //if (res is Double && expected is Double)
        //    is_same = (res.toDouble() - expected.toDouble()).absoluteValue < (precision ?: 0.0)
        //else
        //    is_same = res == expected
        is_same = res == expected
        if (is_same) {
            println("  test #" + nb + ": res=" + res + " CORRECT")
        }
        else {
            println("  test #" + nb + ": res=" + res + " expected=" + expected + " ERROR <-----")
            nb_errors += 1
        }
        nb += 1
    }
    return nb_errors
}

}

File: Notes/Computer Science/Algorithms/Meta/Sebastien Rubens Meta Puzzles Solutions/java/test_all.java
// meta_puzzles by Sebastien Rubens
//
// Please go to https://github.com/seb-pg/meta_puzzles/README.md
// for more information
//
// To the extent possible under law, the person who associated CC0 with
// meta_puzzles has waived all copyright and related or neighboring rights
// to meta_puzzles.
//
// You should have received a copy of the CC0 legalcode along with this
// work.  If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

//trait Result[T]:
//  def get_result(): T

import java.util.function.Function;

public class test_all {

public interface Result<T> {
    T getResult();
}

static <Args extends Result<Ret>, Ret> int run_all_tests(String name, Args[] args_list, Function<Args, Ret> fnc, Double... precision)
{
    System.out.printf("\n%s\n", name);
    int nb_errors = 0;
    int nb = 1;
    for (Args args : args_list)
    {
        var res = fnc.apply(args);
        var expected = args.getResult();
        var is_same = false;
        if (res instanceof Double)
            is_same = Math.abs((double) res - (double) expected) < precision[0];
        else
            is_same = expected.equals(res);
        if (is_same) {
            System.out.printf("  test #%s: res=%s CORRECT\n", nb, String.valueOf(res));
        }
        else {
            System.out.printf("  test #%s: res=%s expected=%s ERROR <-----\n", nb, String.valueOf(res), String.valueOf(expected));
            nb_errors += 1;
        }
        nb += 1;
    }
    return nb_errors;
}

}
